src/main/java/com/snowflake/kafka/connector/internal/BufferThreshold.java

```diff
@@ -0,0 +1,213 @@
+   *
+   * <p>Config parameter: {@link
+   * com.snowflake.kafka.connector.SnowflakeSinkConnectorConfig#BUFFER_COUNT_RECORDS}
+   */
+  private final long bufferKafkaRecordCountThreshold;
+
+  private final long SECOND_TO_MILLIS = TimeUnit.SECONDS.toMillis(1);
+
+  public BufferThreshold(
+      IngestionMethodConfig ingestionMethodConfig,
+      long flushTimeThresholdSeconds,
+      long bufferSizeThresholdBytes,
+      long bufferKafkaRecordCountThreshold) {
+    this.ingestionMethodConfig = ingestionMethodConfig;
+    this.flushTimeThresholdSeconds = flushTimeThresholdSeconds;
+    this.bufferSizeThresholdBytes = bufferSizeThresholdBytes;
+    this.bufferKafkaRecordCountThreshold = bufferKafkaRecordCountThreshold;
+  }
+
+  /**
+   * Returns true if size of current buffer is more than threshold provided (Both in bytes).
+   *
+   * <p>Threshold is config parameter: {@link
+   * com.snowflake.kafka.connector.SnowflakeSinkConnectorConfig#BUFFER_SIZE_BYTES}
+   *
+   * @param currentBufferSizeInBytes current size of buffer in bytes
+   * @return true if bytes threshold has reached.
+   */
+  public boolean isFlushBufferedBytesBased(final long currentBufferSizeInBytes) {
+    return currentBufferSizeInBytes >= bufferSizeThresholdBytes;
+  }
+
+  /**
+   * Returns true if number of ({@link SinkRecord})s in current buffer is more than threshold
+   * provided.
+   *
+   * <p>Threshold is config parameter: {@link
+   * com.snowflake.kafka.connector.SnowflakeSinkConnectorConfig#BUFFER_COUNT_RECORDS}
+   *
+   * @param currentBufferedRecordCount current size of buffer in terms of number of kafka records
+   * @return true if number of kafka threshold has reached in buffer.
+   */
+  public boolean isFlushBufferedRecordCountBased(final long currentBufferedRecordCount) {
+    return currentBufferedRecordCount != 0
+        && currentBufferedRecordCount >= bufferKafkaRecordCountThreshold;
+  }
+
+  /**
+   * If difference between current time and previous flush time is more than threshold return true.
+   *
+   * @param previousFlushTimeStampMs when were previous buffered records were flushed into internal
+   *     stage for snowpipe based implementation or previous buffered records were sent in
+   *     insertRows API of Streaming Snowpipe.
+   * @return true if time based threshold has reached.
+   */
+  public boolean isFlushTimeBased(final long previousFlushTimeStampMs) {
+    final long currentTimeMs = System.currentTimeMillis();
+    return (currentTimeMs - previousFlushTimeStampMs)
+        >= (this.flushTimeThresholdSeconds * SECOND_TO_MILLIS);
+  }
+
+  /** Get flush time threshold in seconds */
+  public long getFlushTimeThresholdSeconds() {
+    return flushTimeThresholdSeconds;
+  }
+
+  /**
+   * Check if provided snowflake kafka connector buffer properties are within permissible values
+   *
+   * @param providedSFConnectorConfig provided by customer
+   * @return true if valid.
+   */
+  public static boolean validateBufferThreshold(
```

Comment:
This is a rather lengthy method, can this call some private helpers to validate specific config options?

Reply:
I thought about it but there wasn't any benefit of doing it. Let me think through more.